### PR TITLE
Add the ability to capture and load snapshots of memory and message history

### PR DESCRIPTION
**Pull request by Joe0**
Related to:
#9 #122 #124 #125 

Primary discussion here:
https://github.com/Torantulino/Auto-GPT/pull/124#issuecomment-1496237527

TODO:
Tests
Code review
Improve code quality

### CODE REVIEW

File: scripts/chat.py [hunk 1]
```diff
@@ -114,10 +114,10 @@ def chat_with_ai(
             )
 
             # Update full message history
-            full_message_history.append(
+            message_history.append(
```
Comment (sberney) [inline scripts/chat.py:117]: Is this a significant or required change?

File: scripts/commands.py [hunk 2]
```diff
@@ -178,14 +177,14 @@ def get_hyperlinks(url):
 
 def commit_memory(string):
     _text = f"""Committing memory with string "{string}" """
-    mem.permanent_memory.append(string)
+    memory.append(string)
```
Comment (sberney) [inline scripts/commands.py:180]: Nice, the stuff in this area makes sense to me.

File: scripts/data_store.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import os
+import shelve
+import memory
+import message_history
+
+class DataStore():
+
+    def persist_message_history(self, id):
+      return True
```
Comment (sberney) [inline scripts/data_store.py:9]: What about return False on everything for the abstract base class? Or if errors are involved, throwing the NotImplementedError?

File: scripts/data_store.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        return True
+
+    def load_message_history(self, id):
+        try:
+            with shelve.open(f"{self.path}{id}/store") as f:
+                global message_history
+                message_history.set_history(f["message_history"])
+            return True
+        except Exception as e:
+            return e
```
Comment (sberney) [inline scripts/data_store.py:45]: This function returns either `True` or `Error`. Both a consistent boolean or an actual exception is better if  you ask me.

File: scripts/main.py [hunk 5]
```diff
@@ -248,6 +252,9 @@ def parse_arguments():
     parser.add_argument('--speak', action='store_true', help='Enable Speak Mode')
     parser.add_argument('--debug', action='store_true', help='Enable Debug Mode')
     parser.add_argument('--gpt3only', action='store_true', help='Enable GPT3.5 Only Mode')
+    parser.add_argument('--enable-snapshots', action='store_true', help='Enable Snapshots')
+    parser.add_argument('--snapshot-path', type=str, default=None, required=False, help='Path for the snapshot directory')
+    parser.add_argument('--snapshot-id', type=str, default=None, required=False, help='ID of the snapshot to load')
```
Comment (sberney) [inline scripts/main.py:257]: This all looks good right here

**Comment by nponeccop** [comment]
Can you rebase it on the current master?

**Pull request closed without merging**